moveit_setup_assistant/src/tools/moveit_config_data.cpp

```diff
@@ -1550,6 +1595,86 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
   return true;
 }
 
+// ******************************************************************************************
+// Extract the package/stack name from an absolute file path
+// Input:  path
+// Output: package name and relative path
+// ******************************************************************************************
+bool MoveItConfigData::extractPackageNameFromPath(const std::string& path, std::string& package_name,
+                                                  std::string& relative_filepath) const
+{
+  // Get the path to urdf, save filename
+  fs::path file_path = path;
+  fs::path file_directory = file_path;
+  file_directory.remove_filename();
+
+  fs::path sub_path;       // holds the directory less one folder
+  fs::path relative_path;  // holds the path after the sub_path
+
+  // Paths for testing if files exist
+  fs::path package_path;
+
+  std::vector<std::string> path_parts;  // holds each folder name in vector
+
+  // Copy path into vector of parts
+  for (fs::path::iterator it = file_directory.begin(); it != file_directory.end(); ++it)
+    path_parts.push_back(it->string());
+
+  bool package_found = false;
+
+  // reduce the generated directoy path's folder count by 1 each loop
+  for (int segment_length = path_parts.size(); segment_length > 0; --segment_length)
+  {
+    // Reset the sub_path
+    sub_path.clear();
+
+    // Create a subpath based on the outer loops length
+    for (int segment_count = 0; segment_count < segment_length; ++segment_count)
+    {
+      sub_path /= path_parts[segment_count];
+
+      // decide if we should remember this directory name because it is topmost, in case it is the package/stack name
+      if (segment_count == segment_length - 1)
```

Comment:
Instead of checking this condition in the loop, you can move the assignment out from the loop:
`package_name = path_parts[segment_length - 1]`